internal/pkg/vcsutil/vcs_util.go

```diff
@@ -0,0 +1,276 @@
+package vcsutil
+
+import (
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/hashicorp/go-hclog"
+	"github.com/pkg/errors"
+)
+
+var (
+	sshRemoteRegexp  *regexp.Regexp
+	httpRemoteRegexp *regexp.Regexp
+)
+
+func init() {
+	// Regex matching http/https remotes, tokenizing the unique components for replacement.
+	sshRemoteRegexp, _ = regexp.Compile(`git@(.*?):(.*\.git)`)                 // Example: git@git.test:testorg/testrepo.git
+	httpRemoteRegexp, _ = regexp.Compile(`http[s]?:\/\/(.*?\..*?)\/(.*\.git)`) // Example: https://git.test/testorg/testrepo.git
+}
+
+type VcsChecker struct {
+	log hclog.Logger
+
+	// path is the local path to the cloned git repo
+	path string
+
+	// remoteName is the name of the remote that corresponds to the url
+	remoteName string
+
+	// internal go-git repo
+	repo *git.Repository
+}
+
+// NewVcsChecker creates a new configured VcsChecker.
+// NOTE: VcsChecker subprocceses to the `git` command. Git must be installed.
+func NewVcsChecker(log hclog.Logger, path string, remoteUrl string) (*VcsChecker, error) {
+	if _, err := exec.LookPath("git"); err != nil {
+		return nil, errors.New("command 'git' not installed")
+	}
+
+	v := &VcsChecker{
+		log:  log,
+		path: path,
+	}
+
+	var err error
+	v.repo, err = git.PlainOpenWithOptions(path, &git.PlainOpenOptions{
+		DetectDotGit: true,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to open git repo at path %s", path)
+	}
+
+	remoteName, err := v.getRemoteName(remoteUrl)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get remote name from url %s", remoteUrl)
+	}
+	v.remoteName = remoteName
+
+	return v, nil
+}
+
+// RepoHasDiff looks for unstaged, staged, and committed (but not pushed)
+// differences between the local VcsChecker.path repo and the specified
+// remote url and branch
+func (v *VcsChecker) RepoHasDiff(remoteBranch string) (bool, error) {
+	gitStatus, err := v.runVcsGitCommand("status", "-s")
+	if len(gitStatus) > 0 {
+		return true, nil
+	}
+
+	diff, err := v.remoteHasDiff(remoteBranch)
+	if err != nil {
+		return false, err
+	}
+
+	return diff, nil
+}
+
+// FileHasDiff checks only the specified file for unstaged, staged, and
+// committed (but not pushed) differences between the local VcsChecker.path
+// repo and the specified remote url and branch
+func (v *VcsChecker) FileHasDiff(remoteUrl string, remoteBranch string, filename string) (bool, error) {
```

Comment:
Feels a bit dirty that the `New` function requests a remote but then you have to respecify a remote for this API. Why not just use the same remote that was specified for New?